SDL 0280 Revisions - Adding new parameter of requires audio support and bluetoothDeviceAddress (#1043)
smartdevicelink/sdl_evolution: proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress.md

To solve this problem, add `bluetoothDeviceAddress` and `requiresAudioSupport` so that a new SDL device can be specified in `RegisterAppInterface`, and prompt the user to connect the device via BT, or the HU will connect the device via BT automatically when connected via USB.

## Proposed solution

The current SDL Java Suite library cancels the transport connection if the `requiresAudioSupport` setting is TRUE and BT A2DP is not connected.
However, with this proposal, by adding a new parameter, the SDL app always establishes the transport connection and is registered without depending on the connection status of BT A2DP. If BT A2DP is not connected, the HU will automatically connect BT using `bluetoothDeviceAddress` or request connection from the user.

### Change the app registration flow

In order for the app to be always registered, the changes in SDL session establishment to app registration flow are shown below.
If the SDL app or HU cannot use this function, it will perform the same operation as before.

> ![registeringApp_flow](../assets/proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress/registeringApp_flow.png)

Reviewer: The Steering Committee did not request this image change in the proposal revisions.  Please revert back to the original `registeringApp_Flow.png`.  If you would like to suggest updating this image, you can comment on the review issue with the suggested change when the revised proposal is brought back into review.  Thank you!

Author: HI, @theresalech -san.
Thank you for your re-review.
I will replace (revert) the image file to previous one and make a commit it.
Could you please confirm it again.
Thank you.

      </param>
      <param name="isSDLAllowed" type="Boolean" mandatory="false">
          <description>Sent by SDL in UpdateDeviceList. 'true' - if device is allowed for PolicyTable Exchange; 'false' - if device is NOT allowed for PolicyTable Exchange </description>
      </param>
+     <param name="bluetoothDeviceAddress" type="String" mandatory="false">
+         <description>Device BT Address</description>
+     </param>
  </struct>
  ...
  <struct name="HMIApplication">
  ...
      <param name="deviceInfo" type="Common.DeviceInfo" mandatory="true">
          <description>The ID, serial number, and transport type by which the named app's device is connected to HU.</description>
      </param>
  ...
      <param name="cloudConnectionStatus" type="Common.CloudConnectionStatus" mandatory="false"></param>
+     <param name="requiresAudioSupport" type="Boolean" mandatory="false">
+         <description>Set whether or not this app requires the use of an audio streaming output device.</description>
+     </param>
  </struct>
```

### MediaStreamingStatus class

> The library supports the new version of the protocol specification. However the app needs to use `MediaStreamingStatus` class before registering if the IVI system uses the earlier version.

Reviewer: Should this be re-worded, per [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-644850433):
```suggestion
If the Java Suite library supports the newer version of the protocol specification, but the IVI system is using a lower version, the app will still need to use the `MediaStreamingStatus` class before attempting to register.
```

> With the changes of the flow, it is necessary to do refactoring the library to move the logic of `MediaStreamingStatus` class.

Reviewer: ```suggestion
With the changes of the flow, it is necessary to do refactoring of the Java Suite library to move the logic of the `MediaStreamingStatus` class.
```

> 

Reviewer: Based on [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-644850433), additional information about the refactoring and details about how to use the `MediaStreamingStatus` class should be added to the proposal.  Can you please add?

Author: Hi, @theresalech -san.
I am very sorry for delay my response.
I modified the proposal documentation and added the image file.
Also, I made a commit them.
Could you please confirm them again.
Thank you.

### Launch the app

By using this function, the HU is able to perform  BT automatic connection based on the `bluetoothDeviceAddress` which is notified when the app that requests audio (`requiresAudioSupport` is set to true) is launched.
It can also display a message that prompts BT connection to the user.

> Note: While the HMI is connecting with the Bluetooth mobile device of the app that is set up `requiresAudioSupport=true`, preventions from activations of the app are needed.

Reviewer: I think this should be re-worded to be more aligned with [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-644911918)
```suggestion
Note: The HMI will be responsible for preventing app activation while the HMI is in the process of connecting to the mobile device's BT for an app that has `requiresAudioSupport=true`. For example, the HMI could show a loading icon for that app, or display an error message to the user upon activation.
```

### MediaStreamingStatus class

If the Java Suite library supports the newer version of the protocol specification, but the IVI system is using a lower version, the app will still need to use the `MediaStreamingStatus` class before attempting to register.

> By using the value of isAudioOutputAvailable() method on MediaStreamingStatus class, the Java Suite library can make a decision whether BT A2DP is connected like below:

Reviewer: ```suggestion
By using the value of `isAudioOutputAvailable()` method on `MediaStreamingStatus` class, the Java Suite library can make a decision whether BT A2DP is connected like below:
```

This proposal requires a major version change.
Since new parameters are added, Core, iOS, Java Suite, RPC, Protocol, and HMI are affected.

> ## Alternatives considered

Reviewer: The Steering Committee's requested revisions did not include removing this section.  Even though we will be using the flow described in the `Proposed solution` section, the flow described here is still an alternative that was considered, so should be kept for visibility.  Please add back lines 174 - 219.

Author: Hi, @theresalech -san.
Thank you for your review.
I will sdd back lines 174-219 and make a commit it.
Could you please confirm it.
Thank you.

Reviewer: @Akihiro-Miyazaki, this does not yet appear to be resolved.

Author: Hi, @theresalech -san.
I am very sorry for delay my response.
I modified the proposal documentation and added the image file.
Also, I made a commit them.
Could you please confirm them again.
Thank you.

## Alternatives considered

> The flow in the Proposed solution section is complicated, so the author has also considered the simplified solution below.

Reviewer: You can update this section to reflect the reason why we decided to move forward with the flow described in the `Proposed solution` section:
```suggestion
While this alternative was considered, the Steering Committee determined this flow does not account for previous versions of the protocol, and therefore is not viable as the primary solution.
```

Author: Hi, @theresalech -san.
Thank you for your review.
I will modify to this sentence after adding back lines 174-219 and make a commit it.
Could you please confirm it again.
Thank you.

Reviewer: @Akihiro-Miyazaki, this does not yet appear to be resolved.

Author: Hi, @theresalech -san.
I am very sorry for delay my response.
I corressponded it and made a commit it.
Could you please confirm it again.
Thank you.

> There is still room for consideration, but I recommend this proposed solution.

Reviewer: ```suggestion
```
Remove in favor of suggested change to line 174.

This proposal requires a major version change.
Since new parameters are added, Core, iOS, Java Suite, RPC, Protocol, and HMI are affected.

## Alternatives considered

While this alternative was considered, the Steering Committee determined this flow does not account for previous versions of the protocol, and therefore is not viable as the primary solution.

### Change the app registration flow

> ![registeringApp_flow1](../assets/proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress/registeringApp_flow1.png)

Reviewer: This image is for the alternative solution, correct?  The Steering Committee didn't request any revisions to the alternative solution.  Can you please advise why you've changed this image?

Author: Hi, @theresalech -san.
Thank you for your prompt reply.
Yes, it is correct.
I only changed the file name about the proposal documentation and image file.
Thank you.

Reviewer: @Akihiro-Miyazaki, I'm having trouble following if the flow depicted in the image is for the proposed solution or the solution described in the `Alternatives considered` section.  Regardless, this image change was not part of the Steering Committee's [agreed upon revisions](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-648864859) and so should not be included in this PR.  Please revert to the image used previously, and if you would like a new image to be included in the proposal, you can comment on the review issue with the suggested change when the revised proposal is brought back into review.  Please let me know if you have any questions about this.  Thank you!

Author: @theresalech -san,
I modified a proposal documentation.
I also replaced the image files.
And I made a commit them.
Thank you.

Reviewer: @Akihiro-Miyazaki , the image file should **not** be replaced in this PR, as that was not requested by the Steering Committee.  Please revert to the image used previously (registeringApp_flow2.png).  Thank you.

Author: @theresalech -san,
Thank you for your response.
What I did only was to replace (revert) the image file to previous one.
If this comment was wrong, could you please teach me how to revert?
Thank you in advance for your help.

Reviewer: @Akihiro-Miyazaki I've confirmed the new image is no longer present in the `Alternatives Considered` section.  Thanks!
